### ISSUE TITLE
Ghost Batch Normalization performance

### ISSUE THREAD

**Issue by MustafaMustafa** [tensorflow/tensorflow#23049]
GBN seems to be at least twice as slow as regular BN. This is the case even when using the full batch for normalization in GBN mode (`virtual_batch_size=batch_size`).

You can reproduce that with tensorflow/models cifar10 example by running the script below, also available at (https://gist.github.com/MustafaMustafa/a12485746e4c877620a818d227982e1c):

```bash

git clone git@github.com:tensorflow/models.git slow_GBN_issue
cd slow_GBN_issue/tutorials/image/cifar10_estimator

# use tf.layer.batch_normalization
sed -i  's/tf.contrib.layers.batch_norm/tf.layers.batch_normalization/' model_base.py
sed -i  's/decay=/momentum=/' model_base.py
sed -i  's/is_training=/training=/' model_base.py
sed -i  '/data_format=data_format)/c\axis=-1 if data_format=="NHWC" else 1)' model_base.py

# make val dataset a multiple of batchsize (multiple of 128)
sed -i  's/return 1000/return 9984/' cifar10.py

python generate_cifar10_tfrecords.py --data-dir=${PWD}/cifar-10-data
if [ -d ./cifar10_logs ]
then
  rm -rf ./cifar10_logs
fi
python3 cifar10_main.py --data-dir=${PWD}/cifar-10-data --job-dir=./cifar10_logs --num-gpus=1 --eval-batch-size 128 --train-steps=500 2>&1 | tee bn.log

# use virtual_batch_size
sed -i '/fused/a virtual_batch_size=128,' model_base.py
if [ -d ./cifar10_logs ]
then
  rm -rf ./cifar10_logs
fi
python3 cifar10_main.py --data-dir=${PWD}/cifar-10-data --job-dir=./cifar10_logs --num-gpus=1 --eval-batch-size 128 --train-steps=500 2>&1 | tee gbn.log

# show samples of timing prints
echo "BN:"
grep "sec)" bn.log | tail -1
echo "GBN:"
grep "sec)" gbn.log | tail -1
```

Output:
```
BN:
INFO:tensorflow:learning_rate = 0.1, loss = 2.09511 (4.351 sec)
GBN:
INFO:tensorflow:learning_rate = 0.1, loss = 2.37345 (9.071 sec)
```
------------------------

### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow version (use command below)**: v1.12.0-rc0-0-g1a6dea3 1.12.0-rc0
- **Python version**: 3.5.2
- **Bazel version (if compiling from source)**: 0.16.0
- **GCC/Compiler version (if compiling from source)**: 7.1.0
- **CUDA/cuDNN version**: 9.2/7.1.4
- **GPU model and memory**: Titan X Pascal 12GB
- **Have I written custom code**: No
- **TensorFlow installed from**: Compiled from source
- **Exact command to reproduce**: bash script for how to reproduce provided in the description.
- **Mobile device**: N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
TensorFlow installed from
Exact command to reproduce
Mobile device

**Comment by tensorflowbutler** [comment]
It has been 21 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by MustafaMustafa** [comment]
This is still an issue in TF 1.12 as well. I am not sure there is any reason the entire network should 2x slower with virtual batches. This is the case even when using `virtual batchsize = batchsize`.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Saduf2019** [comment]
@MustafaMustafa 
This is fixed in tf 2.x, could you please verify and let us know.

**Comment by Saduf2019** [comment]
Moving this to closed status due to lack of activity.
